src/libtriton/arch/x86/x86Semantics.cpp

```diff
@@ -5417,11 +5420,19 @@ namespace triton {
             auto& src  = inst.operands[1];
             auto  op1  = this->symbolicEngine->getOperandAst(inst, dst);
             auto  op2  = this->symbolicEngine->getOperandAst(inst, src);
-            auto  node = this->astCtxt->bvmul(this->astCtxt->sx(dst.getBitSize(), op1), this->astCtxt->sx(src.getBitSize(), op2));
-            auto  expr = this->symbolicEngine->createSymbolicExpression(inst, this->astCtxt->extract(dst.getBitSize()-1, 0, node), dst, "IMUL operation");
+            auto  node1 = this->astCtxt->bvmul(op1, op2);
+            auto  node2 = this->astCtxt->bvmul(this->astCtxt->sx(dst.getBitSize(), op1), this->astCtxt->sx(src.getBitSize(), op2));
+            auto  node3 = (dst.getBitSize() == src.getBitSize()) ? node1 : this->astCtxt->extract(dst.getBitSize()-1, 0, node2);
+            // if(dst.getBitSize() == src.getBitSize()) {
+            //   auto expr = this->symbolicEngine->createSymbolicExpression(inst, node1, dst, "IMUL operation");
+            // }
+            // else {
+            //   auto expr = this->symbolicEngine->createSymbolicExpression(inst, this->astCtxt->extract(dst.getBitSize()-1, 0, node2), dst, "IMUL operation");
+            // }
```

Comment:
Mmmh, why this condition (node3)? Can you provide an example where operands are of different size? Especially where the destination est smaller than the source (this is what you assume according to your extraction).

Reply:
I just added ``node3`` to be consistent with all the other cases. But if it finds out src and dst are necessarily of the same size, the extract of sz is just useless as the semantic does not seems to keep the higher bits.

Reply:
I was not sure, that `src` and `dst` are necessarily of same size that's why I left the condition

Reply:
For this one the semantic is:

    ELSE IF (NumberOfOperands = 2)
        THEN
            TMP_XP ← DEST ∗ SRC (* Signed multiplication; TMP_XP is a signed integer at twice the width of the SRC *)
            DEST ← TruncateToOperandSize(TMP_XP);
            IF SignExtend(DEST) ≠ TMP_XP
                THEN CF ← 1; OF ← 1;
                ELSE CF ← 0; OF ← 0; FI;
Thus the higher part, is truncated and definitely lost. For operand sizes the manual specifies:

> Two-operand form — The source operand is promoted to 64 bits if it is a register or a memory location. The destination operand is promoted to 64 bits.

Thus I assume src and dst are necessarily of the same size. Thus we can in get rid of the `node3`. However `node2` is still required to compute cf and of.